hathor/indexes/rocksdb_height_index.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Optional, Tuple
+
+from structlog import get_logger
+
+from hathor.conf import HathorSettings
+from hathor.indexes.height_index import BLOCK_GENESIS_ENTRY, HeightIndex, IndexEntry
+from hathor.indexes.rocksdb_utils import RocksDBIndexUtils
+
+if TYPE_CHECKING:  # pragma: no cover
+    import rocksdb
+
+settings = HathorSettings()
+logger = get_logger()
+
+_CF_NAME_HEIGHT_INDEX = b'height-index'
+
+
+class RocksDBHeightIndex(HeightIndex, RocksDBIndexUtils):
+    """ Index of blocks by height.
+
+    This index uses the following key/value format:
+
+        key = [height]
+              |--4b--|
+
+        value = [tx.timestamp][tx.hash]
+                |--4 bytes---||--32b--|
+
+    It works nicely because rocksdb uses a tree sorted by key under the hood.
+    """
+
+    def __init__(self, db: 'rocksdb.DB', *, cf_name: Optional[bytes] = None) -> None:
+        RocksDBIndexUtils.__init__(self, db)
+        self.log = logger.new()
+
+        # column family stuff
+        self._cf_name = cf_name or _CF_NAME_HEIGHT_INDEX
+        self._cf = self._fresh_cf(self._cf_name)
+
+        # XXX: when we stop using a fresh column-family we have to figure-out when to not run this
+        self._init_db()
+
+    def _init_db(self) -> None:
+        """ Initialize the database with the genesis entry."""
+        key_genesis = self._to_key(0)
+        value_genesis = self._to_value(BLOCK_GENESIS_ENTRY)
+        self._db.put((self._cf, key_genesis), value_genesis)
+
+    def _to_key(self, height: int) -> bytes:
+        """ Serialize height to key used internally"""
+        import struct
+        key = struct.pack('>I', height)
+        assert len(key) == 4
+        return key
+
+    def _from_key(self, key: bytes) -> int:
+        """ Parse internal key to the height"""
+        import struct
+        assert len(key) == 4
+        (height,) = struct.unpack('>I', key)
+        return height
+
+    def _to_value(self, entry: IndexEntry) -> bytes:
+        """ Serialize entry to the value used internally"""
+        import struct
+        assert len(entry.hash) == 32
+        value = bytearray(entry.hash)
+        value.extend(struct.pack('>I', entry.timestamp))
+        assert len(value) == 32 + 4
```

Comment:
Will this give us `[tx.timestamp][tx.hash]` or `[tx.hash][tx.timestamp]`? The docstring in the beginning says `[tx.timestamp][tx.hash]`.

Reply:
The docstring is wrong. I'm fixing it. It's actually `[tx.hash][tx.timestamp]`, although the order doesn't really matter here.